src/initio/initio.c

```diff
@@ -0,0 +1,852 @@
+        }
+    } else {
+        syslog(LOG_ERR, "parse_pwm: invalid pwm pin number");
+    }
+
+    return dev;
+}
+
+static mraa_iio_context
+parse_iio(char** proto, size_t n)
+{
+    if (proto == NULL || n <= 1) {
+        syslog(LOG_ERR, "parse_iio: invalid parameters");
+        return NULL;
+    }
+
+    mraa_iio_context dev = NULL;
+
+    int device = -1;
+    if (proto[1] && (mraa_atoi_x(proto[1], NULL, &device, 0) == MRAA_SUCCESS)) {
+        dev = mraa_iio_init(device);
+        if (dev == NULL) {
+            syslog(LOG_ERR, "parse_iio: could not init iio device %d", device);
+        }
+    } else {
+        syslog(LOG_ERR, "parse_iio: invalid iio device number");
+    }
+
+    return dev;
+}
+
+static mraa_i2c_context
+parse_i2c(char** proto, size_t n)
+{
+    if (proto == NULL || n <= 1) {
+        syslog(LOG_ERR, "parse_i2c: invalid parameters");
+        return NULL;
+    }
+
+    int idx = 1;
+    mraa_i2c_context dev = NULL;
+
+    int bus = -1;
+    if (proto[idx] && (mraa_atoi_x(proto[idx], NULL, &bus, 0) == MRAA_SUCCESS)) {
+        dev = mraa_i2c_init(bus);
+        if (dev == NULL) {
+            syslog(LOG_ERR, "parse_i2c: could not init i2c bus %d", bus);
+            return NULL;
+        } else {
+            if (++idx == n)
+                return dev;
+        }
+    } else {
+        syslog(LOG_ERR, "parse_i2c: invalid i2c bus number");
+        return NULL;
+    }
+
+    int address = -1;
+    if (proto[idx] && (mraa_atoi_x(proto[idx], NULL, &address, 0) == MRAA_SUCCESS)) {
+        if (mraa_i2c_address(dev, (uint8_t) address) != MRAA_SUCCESS) {
+            syslog(LOG_ERR, "parse_i2c: error setting up i2c address");
+            mraa_i2c_stop(dev);
+            return NULL;
+        } else {
+            if (++idx == n) {
+                return dev;
+            }
+        }
+    }
+
+    int mode = -1;
+    if (strncmp(proto[idx], I_MODE_STD, strlen(I_MODE_STD)) == 0) {
+        mode = MRAA_I2C_STD;
+    } else if (strncmp(proto[idx], I_MODE_FAST, strlen(I_MODE_FAST)) == 0) {
+        mode = MRAA_I2C_FAST;
+    } else if (strncmp(proto[idx], I_MODE_HIGH, strlen(I_MODE_HIGH)) == 0) {
+        mode = MRAA_GPIO_PULLDOWN;
+    }
+
+    if (mode != -1) {
+        if (mraa_i2c_frequency(dev, (mraa_i2c_mode_t) mode) != MRAA_SUCCESS) {
+            syslog(LOG_ERR, "parse_i2c: error setting up gpio driver mode %s", proto[idx]);
+            mraa_i2c_stop(dev);
+            return NULL;
+        }
+    }
+
+    return dev;
+}
+
+static mraa_aio_context
+parse_aio(char** proto, size_t n)
+{
+    if (proto == NULL || n <= 1) {
+        syslog(LOG_ERR, "parse_aio: invalid parameters");
+        return NULL;
+    }
+
+    int idx = 1;
+    mraa_aio_context dev = NULL;
+
+    int aio_num = -1;
+    if (proto[idx] && (mraa_atoi_x(proto[idx], NULL, &aio_num, 0) == MRAA_SUCCESS)) {
+        dev = mraa_aio_init(aio_num);
+        if (dev == NULL) {
+            syslog(LOG_ERR, "parse_aio: could not init aio number %d", aio_num);
```

Comment:
Ideally, all errors would tell the user exactly what a failure is and give them a hint on how to solve it.  

For example:

Could not initialize aio '1' from initialization string 'a:1:out'